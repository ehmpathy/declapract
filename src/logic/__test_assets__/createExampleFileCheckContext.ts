import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
