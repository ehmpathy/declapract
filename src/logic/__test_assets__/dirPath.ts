export const testAssetsDirectoryPath = __dirname;
