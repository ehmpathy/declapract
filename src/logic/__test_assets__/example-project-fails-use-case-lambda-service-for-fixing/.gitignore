# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

