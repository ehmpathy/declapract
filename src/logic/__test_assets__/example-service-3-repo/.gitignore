.declapract
node_modules
package-lock.json
