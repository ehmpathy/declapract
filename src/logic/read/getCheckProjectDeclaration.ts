import { CheckProjectDeclaration } from '../../domain/objects/CheckProjectDeclaration';
import { readFileAsync } from '../../utils/fileio/readFileAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';
import { UnexpectedCodePathError } from '../UnexpectedCodePathError';
import { getCheckFileDeclaration } from './getCheckFileDeclaration/getCheckFileDeclaration';

export const getCheckProjectDeclaration = async ({
  declaredProjectDirectory,
}: {
  declaredProjectDirectory: string;
}): Promise<CheckProjectDeclaration> => {
  // grab name from the directory
  const bestPracticeName = (new RegExp(/\/([\w-]+)\/best-practice$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  const badPracticeName = (new RegExp(/bad-practices\/([\w-]+)$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  const name = bestPracticeName ?? badPracticeName;
  if (!name)
    throw new UnexpectedCodePathError(
      `neither best-practice name nor bad-practice name was extractable from the declared project directory '${declaredProjectDirectory}'`,
    );

  // grab paths to _all_ files in this dir (not just at root level)
  const paths = await listFilesInDirectory(declaredProjectDirectory);

  // grab the meta files (i.e., path matches `${projectRoot}/.declapract.*`)
  const metaFilePaths = paths.filter((path) => new RegExp(/^\.declapract\./).test(path));

  // group all of the other files by main file name (i.e., key = filePath.replace('.declapract.ts$', ''))
  const projectFilePaths = [
    ...new Set(
      paths.filter((path) => !metaFilePaths.includes(path)).map((path) => path.replace(/\.declapract\.ts$/, '')),
    ),
  ];

  // for each "main file", get the CheckFileDefinition, now that we have all the files defined for it
  const checks = await Promise.all(
    projectFilePaths.map((declaredFileCorePath) =>
      getCheckFileDeclaration({ declaredProjectDirectory, declaredFileCorePath }),
    ),
  );

  // get readme contents, if readme defined
  const readme = metaFilePaths.includes('.declapract.readme.md')
    ? await readFileAsync({ filePath: `${declaredProjectDirectory}/.declapract.readme.md` })
    : null;

  // return the project def
  return new CheckProjectDeclaration({
    name,
    dir: declaredProjectDirectory,
    checks,
    readme,
  });
};
