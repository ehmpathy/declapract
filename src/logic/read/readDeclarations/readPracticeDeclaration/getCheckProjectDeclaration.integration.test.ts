import { testAssetsDirectoryPath } from '../../../__test_assets__/dirPath';
import { getCheckProjectDeclaration } from './getCheckProjectDeclaration';

describe('getCheckProjectDeclarationFromDirectory', () => {
  it('should define correctly when only one root level file check and a readme', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times/best-practice`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
  it('should define correctly when many files in many levels', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/terraform/best-practice`,
    });
    expect(project.checks.length).toEqual(7);
    expect(project).toMatchSnapshot();
  });
  it('should define correctly for a bad practice', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times/bad-practices/moment`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
});
