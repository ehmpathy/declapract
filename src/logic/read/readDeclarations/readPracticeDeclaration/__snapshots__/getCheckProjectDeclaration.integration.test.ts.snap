// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCheckProjectDeclarationFromDirectory should define correctly for a bad practice 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "package.json",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "name": "moment",
  "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
}
`;

exports[`getCheckProjectDeclarationFromDirectory should define correctly when many files in many levels 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": ".gitignore",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": ".terraform-version",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "provision/terraform/environments/dev/main.tf",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "provision/terraform/environments/dev/versions.tf",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "provision/terraform/environments/prod/main.tf",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "provision/terraform/environments/prod/versions.tf",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "provision/terraform/product/variables.tf",
      "required": true,
      "type": "EQUALS",
    },
  ],
  "name": "terraform",
  "readme": "best pattern found for terraform so far is to put different environments into their own directory - and just have them reference a shared product dir
",
}
`;

exports[`getCheckProjectDeclarationFromDirectory should define correctly when only one root level file check and a readme 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "package.json",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "name": "dates-and-times",
  "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
}
`;
